src/SortingAlgorithms.jl

```diff
@@ -626,4 +675,399 @@ else
     end
 end
 
+###
+# ThreadedPagedMergeSort
+###
+
+# merge v[lo:hiA] and v[hiA+1:hi] ([A;B])  using buffer t[1:1 + hi-lo]
+function twoended_merge!(v::AbstractVector{T}, t::AbstractVector{T}, lo::Integer, hiA::Integer, hi::Integer, o::Ordering) where T
+    @assert lo <= hiA <= hi
+    loA = lo
+    loB = hiA + 1
+    hiB = hi
+
+    # output array indices
+    oL = 1
+    oR = 1 + hi - lo
+
+    # input array indices
+    iAL = loA
+    iBL = loB
+    iAR = hiA
+    iBR = hiB
```

Comment:
Could you clarify that? I don't think I understand exactly what you mean. I use A and B to avoid confusion between the left subarray and the left index. lo/hi are used to specify the input size/location, not as running indices. Here is a table with the variable names:

| twoended_merge! | merge! | MergeSort (Base.Sort) | Description | 
| ------------- | ------------- | ------------- | ------------- |
| lo | lo  | lo | first index of the input |
| hi | hi | hi | last index of the input |
| hiA |  | m | last index of the left subarray |
| iAL | iA | i | current left index of the left input subarray |
| iAR |  |  | current right index of the left input subarray |
| iBL | iB | j | current left index of the right input subarray |
| iBR |  |  | current right index of the right input subarray |
| oL | k | k | current left index of the output |
| oR |  |  | current right index of the output |

Reply:
The fact that `L` means `lo` but a running index and `R` means `hi` but a running index is unintuitive to me. I don't like switching from one synonym or metaphor to another to mean something unrelated. I see no connection between the low/high metaphor and static variables nor between the left/right metaphor and mutable variables.

As an alternative,

| twoended_merge! | merge! | MergeSort (Base.Sort) | Description | 
| ------------- | ------------- | ------------- | ------------- |
| lo | lo  | lo | first index of the input |
| hi | hi | hi | last index of the input |
| m |  | m | last index of the left subarray |
| a_lo | a | a | current left index of the left input subarray |
| a_hi |  |  | current right index of the left input subarray |
| b_lo | b | b | current left index of the right input subarray |
| b_hi |  |  | current right index of the right input subarray |
| o_lo | k | k | current left index of the output |
| o_hi |  |  | current right index of the output |

I like this because it has one fewer component (lo/hi and a/b/lo rather than lo/hi, iA/iB/o, and L/R); because mutable variables and static variables are more clearly distinguishable (unlike imo `hiA` and `iAR` which blend together); and because the call signature variable names (`lo`/`hi`/`m`) make sense as a triple without needing to dive into the mutable variable name conventions (imo unlike `lo`/`hi`/`hiA`)

Reply:
I used L and R, because in some papers on Quicksort the pointers are called left and right pointers (L, R). And somehow the mutable working indices and the static limits lo/hi were conceptually not the same for me.

But code must be easily understood by people who did not write it, so your opinion is more important. Besides, I like your suggestion and implemented it in [7b2a653](https://github.com/JuliaCollections/SortingAlgorithms.jl/pull/71/commits/7b2a653b24877d464bef9d9d57649a3011050985) with one small change:
`o_hi` is visually easy to confuse with the ordering `o`. And since `i` for input indices is not used anymore, using `o` is not really necessary for consistency, so I use `k` for the output indices everywhere.

Reply:
I like that change. `o` is a bad variable name for integers anyways :)